lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/TaskManager.cs

```diff
@@ -0,0 +1,474 @@
+        /// Closes the IRunningTask 
+        /// Then move the task state to WaitingTaskToClose
+        /// Throw IMRUSystemException if runningTask is null or the running task is already added in the running task collection
+        /// </summary>
+        /// <param name="runningTask"></param>
+        /// <param name="closeMessage"></param>
+        internal void RecordRunningTaskDuringSystemFailure(IRunningTask runningTask, string closeMessage)
+        {
+            if (runningTask == null)
+            {
+                Exceptions.Throw(new IMRUSystemException("RunningTask is null."), Logger);
+            }
+
+            if (_runningTasks.ContainsKey(runningTask.Id))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The task [{0}] is already in running tasks.", runningTask.Id);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            if (TaskState(runningTask.Id) == StateMachine.TaskState.TaskSubmitted)
+            {
+                UpdateState(runningTask.Id, TaskStateEvent.RunningTask);
+            }
+            runningTask.Dispose(Encoding.UTF8.GetBytes(closeMessage));
+            UpdateState(runningTask.Id, TaskStateEvent.WaitingTaskToClose);
+        }
+
+        /// <summary>
+        /// Gets error type based on the information in IFailedTask 
+        /// Currently we use the Message in IFailedTask to distinguish different types of errors
+        /// </summary>
+        /// <param name="failedTask"></param>
+        /// <returns></returns>
+        private TaskStateEvent GetTaskErrorEvent(IFailedTask failedTask)
+        {
+            switch (failedTask.Message)
+            {
+                case TaskAppError:
+                    _numberOfAppError++;
+                    return TaskStateEvent.FailedTaskAppError;
+                case TaskSystemError:
+                    return TaskStateEvent.FailedTaskSystemError;
+                case TaskGroupCommunicationError:
+                    return TaskStateEvent.FailedTaskCommunicationError;
+                default:
+                    return TaskStateEvent.FailedTaskSystemError;
+            }
+        }
+
+        /// <summary>
+        /// Returns the number of application error caused by FailedTask
+        /// </summary>
+        /// <returns></returns>
+        internal int NumberOfAppError()
+        {
+            return _numberOfAppError;
+        }
+
+        /// <summary>
+        /// Checks if all the tasks are in final states
+        /// </summary>
+        /// <returns></returns>
+        internal bool AllInFinalState()
+        {
+            return _tasks.All(t => t.Value.TaskState.IsFinalState());
+        }
+
+        /// <summary>
+        /// Gets current state of the task
+        /// </summary>
+        /// <param name="taskId"></param>
+        /// <returns></returns>
+        internal TaskState TaskState(string taskId)
```

Comment:
`GetTaskState` (so that it looks more like a getter and less like a constructor)